vm/gvisor/gvisor.go

```diff
@@ -117,6 +117,13 @@ func (pool *Pool) Create(workdir string, index int) (vmimpl.Instance, error) {
 	if err := osutil.CopyFile(bin, filepath.Join(imageDir, "init")); err != nil {
 		return nil, err
 	}
+	path := filepath.Join(imageDir, "etc")
+	osutil.MkdirAll(path)
+	hostsData := []byte("127.0.0.1	localhost\n")
```

Comment:
> SYZFAIL: failed to resolve manager addr
> addr=localhost h_errno=2

Why does it try to resolve "localhost"?
Forward returns "stdin:0", which is handled specially in executor.
Can we make instead whatever passes "localhost" to pass "127.0.0.1"? It looks better to me (not adding new code).

Reply:
> Why does it try to resolve "localhost"?

I think I should ask this question. We see a lot of reports in https://syzkaller.appspot.com/bug?extid=de089d67c95c7a5c2f54. In the gVisor log, I see this error:
```
SYZFAIL: failed to resolve manager addr
addr=localhost h_errno=2
(errno 11: Resource temporarily unavailable
```

Reply:
```
I0715 23:59:04.631314       1 strace.go:605] [  18:  18] syz-execprog X execve(0xc001de6bc0 /syz-executor, 0xc001de2db0 ["/syz-executor", "runner", "local", "localhost", "17966"], 0xc001a9e0c0 ["SYZ_GVISOR_PROXY=1", "HOME=/"]) = 0 (0x0) (631.593µs)
```

Reply:
It works with this patch:
```
diff --git a/pkg/rpcserver/local.go b/pkg/rpcserver/local.go
index 1420af606..7b0254099 100644
--- a/pkg/rpcserver/local.go
+++ b/pkg/rpcserver/local.go
@@ -62,7 +62,7 @@ func RunLocal(cfg *LocalConfig) error {
        defer serv.ShutdownInstance(name, true)
 
        bin := cfg.Executor
-       args := []string{"runner", name, "localhost", fmt.Sprint(serv.Port)}
+       args := []string{"runner", name, "127.0.0.1", fmt.Sprint(serv.Port)}
        if cfg.GDB {
                bin = "gdb"
                args = append([]string{
```

Reply:
This happens in tests, right? Then it can come from here:
https://github.com/google/syzkaller/blob/master/pkg/rpcserver/local.go#L65

We actually hit a related problem in another context, where gethostbyname misbehaves and returns only ipv4 address on a machine that supports only ipv6. cc @Mandragorian

Alternatively we can special-case "localhost" and resolve it manually to ipv4 127.0.0.1 and then ipv6 0:0:0:0:0:0:0:1. It should solve both problems.

Reply:
In that other context it's not easy to modify /etc/hosts and we cannot replace "localhost" with 127.0.0.1 b/c a machine may support only ipv6.

Reply:
Do we have any machine that doesn't support ipv6? Can we always connect to ::1?

Reply:
Another option is to get addresses from the loopback device and connect to one of them.

Reply:
> Another option is to get addresses from the loopback device and connect to one of them.

If you are up to writing this in portable across Linux/*BSD/gVisor/Starnix C code :)

> Do we have any machine that doesn't support ipv6? Can we always connect to ::1?

"We" is all current and future syzkaller users. I wouldn't make such assumptions.

It should relatively easy to incorporate parsing of loopback ipv4/6 addresses into this part:
https://github.com/google/syzkaller/blob/master/executor/conn.h#L103-L112

Reply:
Nice, thanks! This should also resolve another problem @Mandragorian observed.